src/Module.zig

```diff
@@ -4339,6 +4362,12 @@ pub fn embedFile(mod: *Module, cur_file: *File, rel_file_path: []const u8) !*Emb
     const bytes = try file.readToEndAllocOptions(gpa, std.math.maxInt(u32), size_usize, 1, 0);
     errdefer gpa.free(bytes);
 
+    const real_path = try std.fs.realpathAlloc(gpa, resolved_path);
+    if (!mem.endsWith(u8, real_path, sub_file_path)) {
+        return EmbedFileResult{ .real_path = real_path };
+    }
+    defer gpa.free(real_path);
```

Comment:
Intuitively I thought the same. But…

We want to successfully load the file first before checking for the path naming issue. If we reverse the order even missing files have to be handled through this error, but this would require a different error message. Reading the file might still go wrong because of some missing privileges. So we would have to do it twice.

Doing the extra work of loading the file should not be an issue here, as most errors will occur only after loading the file during the processing of it.

Reply:
Not sure I understand what you're saying.

> We want to successfully load the file first before checking for the path naming issue. If we reverse the order even missing files have to be handled through this error, but this would require a different error message. Reading the file might still go wrong because of some missing privileges. So we would have to do it twice.

Why? Does `realpath` not report all the same errors as `open`?

Reply:
I don't know. I would guess this is depending on the actual platform and implementation of `real_path` for that platform. There are already multiple version in the zig standard library. I had to choose one that is supported on all compiler platforms. As I don't own all of them I tried to make the least amount of assumptions about them.

Reply:
> Why? Does `realpath` not report all the same errors as `open`?

Looking a bit deeper it seems that different operating systems use different strategies for real path here.
The common platforms like posix and windows open the file and use the file handle to retrieve the real path.
But WASI is currently a very simple emulation using just resolve.
The libc implementation is also free to do whatever it likes to do.

So it seems the file is opened twice for some platforms, because this is the `realpath` is implemented for most platforms (unlike eg. getFdPath).
This concerns me more than the order of operations. But I guess we don't want to optimize on the os level in the compiler code, yet.

Checking after loading the file should no be an issue. If we get an error here, the compilation was fast anyways, as we did not parse and analyze the contents.
So why bother with that?